File: trakt/src/main/java/net/simonvt/trakt/ui/fragment/MovieFragment.java
package net.simonvt.trakt.ui.fragment;

import butterknife.InjectView;
import butterknife.Views;

import com.squareup.otto.Bus;

import net.simonvt.trakt.R;
import net.simonvt.trakt.TraktApp;
import net.simonvt.trakt.event.OnTitleChangedEvent;
import net.simonvt.trakt.provider.TraktContract;
import net.simonvt.trakt.scheduler.MovieTaskScheduler;
import net.simonvt.trakt.ui.dialog.RatingDialog;
import net.simonvt.trakt.widget.RemoteImageView;

import android.database.Cursor;
import android.os.Bundle;
import android.support.v4.app.LoaderManager;
import android.support.v4.content.CursorLoader;
import android.support.v4.content.Loader;
import android.text.format.DateUtils;
import android.view.LayoutInflater;
import android.view.Menu;
import android.view.MenuInflater;
import android.view.MenuItem;
import android.view.View;
import android.view.ViewGroup;
import android.widget.LinearLayout;
import android.widget.RatingBar;
import android.widget.TextView;

import javax.inject.Inject;

public class MovieFragment extends ProgressFragment implements LoaderManager.LoaderCallbacks<Cursor> {

    private static final String TAG = "MovieFragment";

    private static final String ARG_ID = "net.simonvt.trakt.ui.fragment.MovieFragment.id";
    private static final String ARG_TITLE = "net.simonvt.trakt.ui.fragment.MovieFragment.title";

    private static final String DIALOG_RATING = "net.simonvt.trakt.ui.fragment.MovieFragment.ratingDialog";

    private static final int LOADER_MOVIE = 400;
    private static final int LOADER_ACTORS = 401;

    @Inject MovieTaskScheduler mMovieScheduler;
    @Inject Bus mBus;

    @InjectView(R.id.year) TextView mYear;
    @InjectView(R.id.banner) RemoteImageView mBanner;
    @InjectView(R.id.overview) TextView mOverview;
    @InjectView(R.id.isWatched) TextView mIsWatched;
    @InjectView(R.id.inCollection) TextView mCollection;
    @InjectView(R.id.inWatchlist) TextView mWatchlist;
    @InjectView(R.id.ratingContainer) View mRatingContainer;
    @InjectView(R.id.rating) RatingBar mRating;

    @InjectView(R.id.actorsParent) LinearLayout mActorsParent;
    @InjectView(R.id.actors) LinearLayout mActors;

    private long mMovieId;

    private String mMovieTitle;

    private int mCurrentRating;

    private boolean mLoaded;

    private boolean mWatched;

    private boolean mCollected;

    private boolean mInWatchlist;

    public static Bundle getArgs(long movieId, String movieTitle) {
        Bundle args = new Bundle();
        args.putLong(ARG_ID, movieId);
        args.putString(ARG_TITLE, movieTitle);
        return args;
    }

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        TraktApp.inject(getActivity(), this);

        setHasOptionsMenu(true);

        Bundle args = getArguments();
        mMovieId = args.getLong(ARG_ID);
        mMovieTitle = args.getString(ARG_TITLE);
    }

    @Override
    public String getTitle() {
        return mMovieTitle == null ? "" : mMovieTitle;
    }

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {
        return inflater.inflate(R.layout.fragment_movie, container, false);
    }

    @Override
    public void onViewCreated(View view, Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);
        Views.inject(this, view);

        mRatingContainer.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                RatingDialog.newInstance(RatingDialog.Type.MOVIE, mMovieId, mCurrentRating)
                        .show(getFragmentManager(), DIALOG_RATING);
            }
        });

        getLoaderManager().initLoader(LOADER_MOVIE, null, this);
        getLoaderManager().initLoader(LOADER_ACTORS, null, mActorsLoader);
    }

    @Override
    public void onDestroyView() {
        if (getActivity().isFinishing() || isRemoving()) {
            getLoaderManager().destroyLoader(LOADER_MOVIE);
            getLoaderManager().destroyLoader(LOADER_ACTORS);
        }
        super.onDestroyView();
    }

    @Override
    public void onCreateOptionsMenu(Menu menu, MenuInflater inflater) {
        if (mLoaded) {
            if (mWatched) {
                menu.add(0, R.id.action_unwatched, 1, R.string.action_unwatched);
            } else {
                menu.add(0, R.id.action_watched, 2, R.string.action_watched);
                if (mInWatchlist) {
                    menu.add(0, R.id.action_watchlist_remove, 5, R.string.action_watchlist_remove);
                } else {
                    menu.add(0, R.id.action_watchlist_add, 6, R.string.action_watchlist_add);
                }
            }

            if (mCollected) {
                menu.add(0, R.id.action_collection_remove, 3, R.string.action_collection_remove);
            } else {
                menu.add(0, R.id.action_collection_add, 4, R.string.action_collection_add);
            }
        }
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        switch (item.getItemId()) {
            case R.id.action_watched:
                mMovieScheduler.setWatched(mMovieId, true);
                return true;

            case R.id.action_unwatched:
                mMovieScheduler.setWatched(mMovieId, false);
                return true;

            case R.id.action_watchlist_add:
                mMovieScheduler.setIsInWatchlist(mMovieId, true);
                return true;

            case R.id.action_watchlist_remove:
                mMovieScheduler.setIsInWatchlist(mMovieId, false);
                return true;

            case R.id.action_collection_add:
                mMovieScheduler.setIsInCollection(mMovieId, true);
                return true;

            case R.id.action_collection_remove:
                mMovieScheduler.setIsInCollection(mMovieId, false);
                return true;
        }

        return false;
    }

    private void updateView(final Cursor cursor) {
        if (cursor == null || !cursor.moveToFirst()) return;
        mLoaded = true;

        final String title = cursor.getString(cursor.getColumnIndex(TraktContract.Movies.TITLE));
        if (!title.equals(mMovieTitle)) {
            mMovieTitle = title;
            mBus.post(new OnTitleChangedEvent());
        }
        final int year = cursor.getInt(cursor.getColumnIndex(TraktContract.Movies.YEAR));
        final String fanart = cursor.getString(cursor.getColumnIndex(TraktContract.Movies.FANART));
        if (fanart != null) {
            mBanner.setImage(fanart);
        }
        mCurrentRating = cursor.getInt(cursor.getColumnIndex(TraktContract.Movies.RATING));
        mRating.setRating(mCurrentRating);

        final String overview = cursor.getString(cursor.getColumnIndex(TraktContract.Movies.OVERVIEW));
        mWatched = cursor.getInt(cursor.getColumnIndex(TraktContract.Movies.WATCHED)) == 1;
        mCollected = cursor.getInt(cursor.getColumnIndex(TraktContract.Movies.IN_COLLECTION)) == 1;
        mInWatchlist = cursor.getInt(cursor.getColumnIndex(TraktContract.Movies.IN_WATCHLIST)) == 1;

        mIsWatched.setVisibility(mWatched ? View.VISIBLE : View.GONE);
        mCollection.setVisibility(mCollected ? View.VISIBLE : View.GONE);
        mWatchlist.setVisibility(mInWatchlist ? View.VISIBLE : View.GONE);

        if (mYear != null) mYear.setText(String.valueOf(year));
        mOverview.setText(overview);

        setContentVisible(true);
        getActivity().invalidateOptionsMenu();
    }

    @Override
    public Loader<Cursor> onCreateLoader(int i, Bundle bundle) {
        CursorLoader loader =
                new CursorLoader(getActivity(), TraktContract.Movies.buildMovieUri(mMovieId), null, null, null, null);
        loader.setUpdateThrottle(2 * DateUtils.SECOND_IN_MILLIS);
        return loader;
    }

    @Override
    public void onLoadFinished(Loader<Cursor> cursorLoader, Cursor cursor) {
        updateView(cursor);
    }

    @Override
    public void onLoaderReset(Loader<Cursor> cursorLoader) {
    }

    private void updateActors(Cursor c) {
        mActorsParent.setVisibility(c.getCount() > 0 ? View.VISIBLE : View.GONE);
        mActors.removeAllViews();

        while (c.moveToNext()) {
            View v = LayoutInflater.from(getActivity()).inflate(R.layout.person, mActors, false);

            RemoteImageView headshot = (RemoteImageView) v.findViewById(R.id.headshot);
            headshot.setImage(c.getString(c.getColumnIndex(TraktContract.MovieActors.HEADSHOT)));
            TextView name = (TextView) v.findViewById(R.id.name);
            name.setText(c.getString(c.getColumnIndex(TraktContract.MovieActors.NAME)));
            TextView character = (TextView) v.findViewById(R.id.job);
            character.setText(c.getString(c.getColumnIndex(TraktContract.MovieActors.CHARACTER)));

            mActors.addView(v);
        }
    }

    private LoaderManager.LoaderCallbacks<Cursor> mActorsLoader = new LoaderManager.LoaderCallbacks<Cursor>() {
        @Override
        public Loader<Cursor> onCreateLoader(int i, Bundle bundle) {
            CursorLoader loader =
                    new CursorLoader(getActivity(), TraktContract.MovieActors.buildFromMovieId(mMovieId), null, null,
                            null, null);
            loader.setUpdateThrottle(2 * DateUtils.SECOND_IN_MILLIS);
            return loader;
        }

        @Override
        public void onLoadFinished(Loader<Cursor> cursorLoader, Cursor cursor) {
            updateActors(cursor);
        }

        @Override
        public void onLoaderReset(Loader<Cursor> cursorLoader) {
        }
    };
}

File: cathode/src/main/java/net/simonvt/cathode/ui/movie/MovieFragment.kt
/*
 * Copyright (C) 2016 Simon Vig Therkildsen
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package net.simonvt.cathode.ui.movie

import android.content.Context
import android.os.Bundle
import android.text.TextUtils
import android.view.LayoutInflater
import android.view.MenuItem
import android.view.View
import android.view.ViewGroup
import android.widget.TextView
import androidx.appcompat.widget.Toolbar
import androidx.fragment.app.viewModels
import androidx.lifecycle.Observer
import androidx.vectordrawable.graphics.drawable.VectorDrawableCompat
import net.simonvt.cathode.R
import net.simonvt.cathode.api.enumeration.ItemType
import net.simonvt.cathode.api.util.TraktUtils
import net.simonvt.cathode.common.ui.fragment.RefreshableAppBarFragment
import net.simonvt.cathode.common.ui.instantiate
import net.simonvt.cathode.common.util.DateStringUtils
import net.simonvt.cathode.common.util.Ids
import net.simonvt.cathode.common.util.Intents
import net.simonvt.cathode.common.util.Joiner
import net.simonvt.cathode.common.util.guava.Preconditions
import net.simonvt.cathode.common.widget.CircleTransformation
import net.simonvt.cathode.common.widget.RemoteImageView
import net.simonvt.cathode.databinding.FragmentMovieBinding
import net.simonvt.cathode.entity.CastMember
import net.simonvt.cathode.entity.Comment
import net.simonvt.cathode.entity.Movie
import net.simonvt.cathode.images.ImageType
import net.simonvt.cathode.images.ImageUri
import net.simonvt.cathode.settings.TraktLinkSettings
import net.simonvt.cathode.sync.scheduler.MovieTaskScheduler
import net.simonvt.cathode.ui.CathodeViewModelFactory
import net.simonvt.cathode.ui.NavigationListener
import net.simonvt.cathode.ui.comments.LinearCommentsAdapter
import net.simonvt.cathode.ui.dialog.CheckInDialog
import net.simonvt.cathode.ui.dialog.CheckInDialog.Type
import net.simonvt.cathode.ui.dialog.RatingDialog
import net.simonvt.cathode.ui.history.AddToHistoryDialog
import net.simonvt.cathode.ui.history.RemoveFromHistoryDialog
import net.simonvt.cathode.ui.lists.ListsDialog
import net.simonvt.cathode.widget.CheckInDrawable
import java.util.Locale
import javax.inject.Inject

class MovieFragment @Inject constructor(
  private val viewModelFactory: CathodeViewModelFactory,
  private val movieScheduler: MovieTaskScheduler
) : RefreshableAppBarFragment() {

  private val viewModel: MovieViewModel by viewModels { viewModelFactory }

  private var movie: Movie? = null
  private var genres: List<String>? = null
  private var cast: List<CastMember>? = null
  private var userComments: List<Comment>? = null
  private var comments: List<Comment>? = null
  private var related: List<Movie>? = null

  private var _binding: FragmentMovieBinding? = null
  private val binding get() = _binding!!

  var movieId: Long = 0
    private set

  private var movieTitle: String? = null

  private var movieOverview: String? = null

  private var currentRating: Int = 0

  private var loaded: Boolean = false

  private var watched: Boolean = false

  private var collected: Boolean = false

  private var inWatchlist: Boolean = false

  private var watching: Boolean = false

  private var checkedIn: Boolean = false

  lateinit var navigationListener: NavigationListener

  private var checkInDrawable: CheckInDrawable? = null

  override fun onAttach(context: Context) {
    super.onAttach(context)
    navigationListener = requireActivity() as NavigationListener
  }

  override fun onCreate(inState: Bundle?) {
    super.onCreate(inState)
    val args = arguments
    movieId = args!!.getLong(ARG_ID)
    movieTitle = args.getString(ARG_TITLE)
    movieOverview = args.getString(ARG_OVERVIEW)

    setTitle(movieTitle)

    viewModel.setMovieId(movieId)
    viewModel.loading.observe(this, Observer { loading -> setRefreshing(loading) })
    viewModel.movie.observe(this, Observer { movie ->
      this@MovieFragment.movie = movie
      updateView()
    })
    viewModel.genres.observe(this, Observer { genres ->
      this@MovieFragment.genres = genres
      updateGenreViews()
    })
    viewModel.cast.observe(this, Observer { castMembers ->
      cast = castMembers
      updateCast()
    })
    viewModel.userComments.observe(this, Observer { userComments ->
      this@MovieFragment.userComments = userComments
      updateComments()
    })
    viewModel.comments.observe(this, Observer { comments ->
      this@MovieFragment.comments = comments
      updateComments()
    })
    viewModel.relatedMovies.observe(this, Observer { movies ->
      related = movies
      updateRelatedView()
    })
  }

  override fun createView(inflater: LayoutInflater, container: ViewGroup?, inState: Bundle?): View {
    _binding = FragmentMovieBinding.inflate(inflater, container, false)
    return binding.root
  }

  override fun onViewCreated(view: View, inState: Bundle?) {
    super.onViewCreated(view, inState)
    val linkDrawable = VectorDrawableCompat.create(resources, R.drawable.ic_link_black_24dp, null)
    binding.content.website.setCompoundDrawablesWithIntrinsicBounds(linkDrawable, null, null, null)
    binding.content.viewOnTrakt.setCompoundDrawablesWithIntrinsicBounds(
      linkDrawable,
      null,
      null,
      null
    )
    binding.content.viewOnImdb.setCompoundDrawablesWithIntrinsicBounds(
      linkDrawable,
      null,
      null,
      null
    )
    binding.content.viewOnTmdb.setCompoundDrawablesWithIntrinsicBounds(
      linkDrawable,
      null,
      null,
      null
    )

    val playDrawable =
      VectorDrawableCompat.create(resources, R.drawable.ic_play_arrow_black_24dp, null)
    binding.content.trailer.setCompoundDrawablesWithIntrinsicBounds(playDrawable, null, null, null)

    binding.content.overview.text = movieOverview

    if (TraktLinkSettings.isLinked(requireContext())) {
      binding.top.rating.setOnClickListener {
        parentFragmentManager.instantiate(
          RatingDialog::class.java,
          RatingDialog.getArgs(RatingDialog.Type.MOVIE, movieId, currentRating)
        ).show(parentFragmentManager, DIALOG_RATING)
      }
    }

    binding.content.cast.castHeader.setOnClickListener(displayCastClickListener)
    binding.content.comments.commentsHeader.setOnClickListener(displayCommentsClickListener)
    binding.content.related.relatedHeader.setOnClickListener(displayRelatedClickListener)

    updateView()
    updateGenreViews()
    updateCast()
    updateRelatedView()
    updateComments()
  }

  override fun onDestroyView() {
    _binding = null
    super.onDestroyView()
  }

  private val displayCastClickListener = View.OnClickListener {
    navigationListener.onDisplayCredits(ItemType.MOVIE, movieId, movieTitle)
  }

  private val displayCommentsClickListener = View.OnClickListener {
    navigationListener.onDisplayComments(ItemType.MOVIE, movieId)
  }

  private val displayRelatedClickListener = View.OnClickListener {
    navigationListener.onDisplayRelatedMovies(movieId, movieTitle)
  }

  override fun onRefresh() {
    viewModel.refresh()
  }

  override fun createMenu(toolbar: Toolbar) {
    if (loaded) {
      val menu = toolbar.menu

      if (checkInDrawable == null) {
        checkInDrawable = CheckInDrawable(toolbar.context)
        checkInDrawable!!.setWatching(watching || checkedIn)
        checkInDrawable!!.setId(movieId)
      }

      val checkInItem: MenuItem

      if (watching || checkedIn) {
        checkInItem = menu.add(0, R.id.action_checkin, 1, R.string.action_checkin_cancel)
      } else {
        checkInItem = menu.add(0, R.id.action_checkin, 1, R.string.action_checkin)
      }

      checkInItem.setIcon(checkInDrawable).setShowAsAction(MenuItem.SHOW_AS_ACTION_IF_ROOM)
      checkInItem.isEnabled = !watching

      menu.add(0, R.id.action_history_add, 3, R.string.action_history_add)

      if (watched) {
        menu.add(0, R.id.action_history_remove, 4, R.string.action_history_remove)
      } else {
        if (inWatchlist) {
          menu.add(0, R.id.action_watchlist_remove, 7, R.string.action_watchlist_remove)
        } else {
          menu.add(0, R.id.action_watchlist_add, 8, R.string.action_watchlist_add)
        }
      }

      if (collected) {
        menu.add(0, R.id.action_collection_remove, 5, R.string.action_collection_remove)
      } else {
        menu.add(0, R.id.action_collection_add, 6, R.string.action_collection_add)
      }

      menu.add(0, R.id.action_list_add, 9, R.string.action_list_add)
    }
  }

  override fun onMenuItemClick(item: MenuItem): Boolean {
    when (item.itemId) {
      R.id.action_history_add -> {
        parentFragmentManager.instantiate(
          AddToHistoryDialog::class.java,
          AddToHistoryDialog.getArgs(AddToHistoryDialog.Type.MOVIE, movieId, movieTitle)
        ).show(parentFragmentManager, AddToHistoryDialog.TAG)
        return true
      }

      R.id.action_history_remove -> {
        if (TraktLinkSettings.isLinked(requireContext())) {
          parentFragmentManager.instantiate(
            RemoveFromHistoryDialog::class.java,
            RemoveFromHistoryDialog.getArgs(
              RemoveFromHistoryDialog.Type.MOVIE,
              movieId,
              movieTitle,
              null
            )
          ).show(parentFragmentManager, RemoveFromHistoryDialog.TAG)
        } else {
          movieScheduler.removeFromHistory(movieId)
        }
        return true
      }

      R.id.action_checkin -> {
        if (!watching) {
          if (checkedIn) {
            movieScheduler.cancelCheckin()
            if (checkInDrawable != null) {
              checkInDrawable!!.setWatching(false)
            }
          } else {
            if (!CheckInDialog.showDialogIfNecessary(
                requireActivity(), Type.MOVIE, movieTitle,
                movieId
              )
            ) {
              movieScheduler.checkin(movieId, null, false, false, false)
              checkInDrawable!!.setWatching(true)
            }
          }
        }
        return true
      }

      R.id.action_checkin_cancel -> {
        movieScheduler.cancelCheckin()
        return true
      }

      R.id.action_watchlist_add -> {
        movieScheduler.setIsInWatchlist(movieId, true)
        return true
      }

      R.id.action_watchlist_remove -> {
        movieScheduler.setIsInWatchlist(movieId, false)
        return true
      }

      R.id.action_collection_add -> {
        movieScheduler.setIsInCollection(movieId, true)
        return true
      }

      R.id.action_collection_remove -> {
        movieScheduler.setIsInCollection(movieId, false)
        return true
      }

      R.id.action_list_add -> {
        parentFragmentManager.instantiate(
          ListsDialog::class.java,
          ListsDialog.getArgs(ItemType.MOVIE, movieId)
        ).show(parentFragmentManager, DIALOG_LISTS_ADD)
        return true
      }
    }

    return super.onMenuItemClick(item)
  }

  private fun updateView() {
    if (view == null || movie == null) {
      return
    }

    loaded = true

    val traktId = movie!!.traktId
    val title = movie!!.title
    if (title != movieTitle) {
      movieTitle = title
      setTitle(movieTitle)
    }

    val backdropUri = ImageUri.create(ImageUri.ITEM_MOVIE, ImageType.BACKDROP, movieId)
    setBackdrop(backdropUri, true)

    currentRating = movie!!.userRating
    binding.top.rating.setValue(movie!!.rating)

    movieOverview = movie!!.overview
    watched = movie!!.watched
    collected = movie!!.inCollection
    inWatchlist = movie!!.inWatchlist
    watching = movie!!.watching
    checkedIn = movie!!.checkedIn

    if (checkInDrawable != null) {
      checkInDrawable!!.setWatching(watching || checkedIn)
    }

    val hasCheckmark = watched || collected || inWatchlist
    binding.content.checkmarks.checkmarks.visibility = if (hasCheckmark) View.VISIBLE else View.GONE
    binding.content.checkmarks.isWatched.visibility = if (watched) View.VISIBLE else View.GONE
    binding.content.checkmarks.inCollection.visibility = if (collected) View.VISIBLE else View.GONE
    binding.content.checkmarks.inWatchlist.visibility = if (inWatchlist) View.VISIBLE else View.GONE

    var infoOneText = ""
    if (movie?.year ?: 0 > 0) {
      infoOneText = movie!!.year.toString()
      if (!TextUtils.isEmpty(movie!!.certification)) {
        infoOneText += ", " + movie!!.certification
      }
    } else {
      if (!movie!!.certification.isNullOrEmpty()) {
        infoOneText = movie!!.certification!!
      }
    }
    binding.top.info1.text = infoOneText

    if (movie?.runtime ?: 0 > 0) {
      val runtime = DateStringUtils.getRuntimeString(requireContext(), movie!!.runtime.toLong())
      binding.top.info2.text = runtime
    }

    binding.content.overview.text = movieOverview

    val trailer = movie!!.trailer
    if (!TextUtils.isEmpty(trailer)) {
      binding.content.trailer.visibility = View.VISIBLE
      binding.content.trailer.setOnClickListener { Intents.openUrl(requireContext(), trailer) }
    } else {
      binding.content.trailer.visibility = View.GONE
    }

    val website = movie!!.homepage
    if (!TextUtils.isEmpty(website)) {
      binding.content.website.visibility = View.VISIBLE
      binding.content.website.text = website
      binding.content.website.setOnClickListener { Intents.openUrl(requireContext(), website) }
    } else {
      binding.content.website.visibility = View.GONE
    }

    val imdbId = movie!!.imdbId
    val tmdbId = movie!!.tmdbId

    binding.content.viewOnTrakt.setOnClickListener {
      Intents.openUrl(
        requireContext(),
        TraktUtils.getTraktMovieUrl(traktId)
      )
    }

    if (!imdbId.isNullOrEmpty()) {
      binding.content.viewOnImdb.visibility = View.VISIBLE
      binding.content.viewOnImdb.setOnClickListener {
        Intents.openUrl(
          requireContext(),
          TraktUtils.getImdbUrl(imdbId)
        )
      }
    } else {
      binding.content.viewOnImdb.visibility = View.GONE
    }

    if (tmdbId > 0) {
      binding.content.viewOnTmdb.visibility = View.VISIBLE
      binding.content.viewOnTmdb.setOnClickListener {
        Intents.openUrl(
          requireContext(),
          TraktUtils.getTmdbMovieUrl(tmdbId)
        )
      }
    } else {
      binding.content.viewOnTmdb.visibility = View.GONE
    }

    invalidateMenu()
  }

  private fun updateGenreViews() {
    if (view == null) {
      return
    }

    if (genres == null || genres!!.isEmpty()) {
      binding.content.genresTitle.visibility = View.GONE
      binding.content.genresTitle.visibility = View.GONE
      binding.content.genres.visibility = View.GONE
    } else {
      val joinedGenres = Joiner.on(", ").join(genres)

      binding.content.genresTitle.visibility = View.VISIBLE
      binding.content.genresTitle.visibility = View.VISIBLE
      binding.content.genres.visibility = View.VISIBLE
      binding.content.genres.text = joinedGenres
    }
  }

  private fun updateCast() {
    if (view == null) {
      return
    }

    binding.content.cast.container.container.removeAllViews()
    if (cast.isNullOrEmpty()) {
      binding.content.cast.cast.visibility = View.GONE
    } else {
      binding.content.cast.cast.visibility = View.VISIBLE

      for (castMember in cast!!) {
        val v = LayoutInflater.from(requireContext()).inflate(
          R.layout.section_people_item,
          binding.content.cast.container.container,
          false
        )

        val headshotUri =
          ImageUri.create(ImageUri.ITEM_PERSON, ImageType.PROFILE, castMember.person.id)

        val headshot = v.findViewById<RemoteImageView>(R.id.headshot)
        headshot.addTransformation(CircleTransformation())
        headshot.setImage(headshotUri)

        val name = v.findViewById<TextView>(R.id.person_name)
        name.text = castMember.person.name

        val character = v.findViewById<TextView>(R.id.person_job)
        character.text = castMember.character

        v.setOnClickListener { navigationListener.onDisplayPerson(castMember.person.id) }

        binding.content.cast.container.container.addView(v)
      }
    }
  }

  private fun updateRelatedView() {
    if (view == null) {
      return
    }

    binding.content.related.container.container.removeAllViews()
    if (related.isNullOrEmpty()) {
      binding.content.related.related.visibility = View.GONE
    } else {
      binding.content.related.related.visibility = View.VISIBLE

      for (movie in related!!) {
        val v = LayoutInflater.from(requireContext()).inflate(
          R.layout.section_related_item,
          binding.content.related.container.container,
          false
        )

        val poster = ImageUri.create(ImageUri.ITEM_MOVIE, ImageType.POSTER, movie.id)

        val posterView = v.findViewById<RemoteImageView>(R.id.related_poster)
        posterView.addTransformation(CircleTransformation())
        posterView.setImage(poster)

        val titleView = v.findViewById<TextView>(R.id.related_title)
        titleView.text = movie.title

        val formattedRating = String.format(Locale.getDefault(), "%.1f", movie.rating)

        val ratingText: String
        if (movie.votes >= 1000) {
          val convertedVotes = movie.votes / 1000.0f
          val formattedVotes = String.format(Locale.getDefault(), "%.1f", convertedVotes)
          ratingText = getString(R.string.related_rating_thousands, formattedRating, formattedVotes)
        } else {
          ratingText = getString(R.string.related_rating, formattedRating, movie.votes)
        }

        val ratingView = v.findViewById<TextView>(R.id.related_rating)
        ratingView.text = ratingText

        v.setOnClickListener {
          navigationListener.onDisplayMovie(movie.id, movie.title, movie.overview)
        }
        binding.content.related.container.container.addView(v)
      }
    }
  }

  private fun updateComments() {
    if (view == null) {
      return
    }

    LinearCommentsAdapter.updateComments(
      requireContext(),
      binding.content.comments.container.container,
      userComments,
      comments
    )
    binding.content.comments.comments.visibility = View.VISIBLE
  }

  companion object {

    private const val TAG = "net.simonvt.cathode.ui.movie.MovieFragment"

    private const val ARG_ID = "net.simonvt.cathode.ui.movie.MovieFragment.id"
    private const val ARG_TITLE = "net.simonvt.cathode.ui.movie.MovieFragment.title"
    private const val ARG_OVERVIEW = "net.simonvt.cathode.ui.movie.MovieFragment.overview"

    private const val DIALOG_RATING = "net.simonvt.cathode.ui.movie.MovieFragment.ratingDialog"
    private const val DIALOG_LISTS_ADD = "net.simonvt.cathode.ui.movie.MovieFragment.listsAddDialog"

    @JvmStatic
    fun getTag(movieId: Long): String {
      return TAG + "/" + movieId + "/" + Ids.newId()
    }

    @JvmStatic
    fun getArgs(movieId: Long, movieTitle: String?, overview: String?): Bundle {
      Preconditions.checkArgument(movieId >= 0, "movieId must be >= 0")

      val args = Bundle()
      args.putLong(ARG_ID, movieId)
      args.putString(ARG_TITLE, movieTitle)
      args.putString(ARG_OVERVIEW, overview)
      return args
    }
  }
}
